### PR TITLE
Rethrow OpenSearch exception in concurrent search

### LINKED ISSUE #9124: [Concurrent Segment Search] Fix Circuit Breaker related ITs
Subtask for https://github.com/opensearch-project/OpenSearch/issues/7357 to focus on CB failures

- org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testMemoryBreaker
```
2023-07-26T13:52:14.1470413Z   2> REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testMemoryBreaker" -Dtests.seed=41D55BF4F404C7FC -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es -Dtests.timezone=Antarctica/Macquarie -Druntime.java=20
2023-07-26T13:52:14.1471308Z   2> java.lang.AssertionError: 
2023-07-26T13:52:14.1471585Z     Expected: <INTERNAL_SERVER_ERROR>
2023-07-26T13:52:14.1471887Z          but: was <TOO_MANY_REQUESTS>
2023-07-26T13:52:14.1472222Z         at __randomizedtesting.SeedInfo.seed([41D55BF4F404C7FC:52BB2E5D942C2472]:0)
2023-07-26T13:52:14.1472664Z         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
2023-07-26T13:52:14.1473112Z         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
2023-07-26T13:52:14.1473800Z         at org.opensearch.test.hamcrest.OpenSearchAssertions.assertFailures(OpenSearchAssertions.java:362)
2023-07-26T13:52:14.1474584Z         at org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testMemoryBreaker(CircuitBreakerServiceIT.java:175)
```
- org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testRamAccountingTermsEnum
```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testRamAccountingTermsEnum" -Dtests.seed=41D55BF4F404C7FC -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es -Dtests.timezone=Antarctica/Macquarie -Druntime.java=20
2023-07-26T13:52:12.8125014Z 
2023-07-26T13:52:12.8125792Z org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT > testRamAccountingTermsEnum FAILED
2023-07-26T13:52:12.8126273Z     java.lang.AssertionError: 
2023-07-26T13:52:12.8126554Z     Expected: <INTERNAL_SERVER_ERROR>
2023-07-26T13:52:12.8126807Z          but: was <TOO_MANY_REQUESTS>
2023-07-26T13:52:12.8127174Z         at __randomizedtesting.SeedInfo.seed([41D55BF4F404C7FC:CB932CD95149F85E]:0)
2023-07-26T13:52:12.8127621Z         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
2023-07-26T13:52:12.8128072Z         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
2023-07-26T13:52:12.8128647Z         at org.opensearch.test.hamcrest.OpenSearchAssertions.assertFailures(OpenSearchAssertions.java:362)
2023-07-26T13:52:12.8129495Z         at org.opensearch.indices.memory.breaker.CircuitBreakerServiceIT.testRamAccountingTermsEnum(CircuitBreakerServiceIT.java:232)
```

### PR DESCRIPTION
**Pull request by neetikasinghal**
### Description

CircuitBreaker tests are breaking because of the difference in the Exceptions being thrown at the shard failure. The exceptions can be found below:

Shard failure exception with Concurrent Search
```
shard[[wkVu-9P-R1ylmYdjXkGw4w][cb-test][6]], reason [RemoteTransportException[[node_t0][127.0.0.1:62392][indices:data/read/search[phase/query/id]]]; nested: QueryPhaseExecutionException[Query Failed [Failed to execute main query]]; nested: CircuitBreakingException[[fielddata] Data too large, data for [test] would be [1689/1.6kb], which is larger than the limit of [100/100b]]; ], cause [CircuitBreakingException[[fielddata] Data too large, data for [test] would be [1689/1.6kb], which is larger than the limit of [100/100b]]
	at org.opensearch.common.breaker.ChildMemoryCircuitBreaker.circuitBreak(ChildMemoryCircuitBreaker.java:104)
	at org.opensearch.common.breaker.ChildMemoryCircuitBreaker.limit(ChildMemoryCircuitBreaker.java:196)
	at ...
```

Shard failure exception without Concurrent Search
```
shard [[cbO2S4OXRHOvPGDErQe2lg][cb-test][0]], reason [RemoteTransportException[[node_t0][127.0.0.1:64587][indices:data/read/search[phase/query]]]; nested: QueryPhaseExecutionException[Query Failed [Failed to execute main query]]; nested: OpenSearchException[java.util.concurrent.ExecutionException: CircuitBreakingException[[fielddata] Data too large, data for [test] would be [3529/3.4kb], which is larger than the limit of [100/100b]]]; nested: ExecutionException[CircuitBreakingException[[fielddata] Data too large, data for [test] would be [3529/3.4kb], which is larger than the limit of [100/100b]]]; nested: CircuitBreakingException[[fielddata] Data too large, data for [test] would be [3529/3.4kb], which is larger than the limit of [100/100b]]; ], cause [OpenSearchException[java.util.concurrent.ExecutionException: CircuitBreakingException[[fielddata] Data too large, data for [test] would be [3529/3.4kb], which is larger than the limit of [100/100b]]]; nested: ExecutionException[CircuitBreakingException[[fielddata] Data too large, data for [test] would be [3529/3.4kb], which is larger than the limit of [100/100b]]]; nested: CircuitBreakingException[[fielddata] Data too large, data for [test] would be [3529/3.4kb], which is larger than the limit of [100/100b]];
	at org.opensearch.index.fielddata.plain.AbstractIndexOrdinalsFieldData.load(AbstractIndexOrdinalsFieldData.java:116)
	at org.opensearch.index.fielddata.plain.AbstractIndexOrdinalsFieldData.load(AbstractIndexOrdinalsFieldData.java:62)
	at ...
```
This PR attempts to fix this.
All tests report after this change(making sure that the other tests don't break with this change): [link](https://github.com/neetikasinghal/OpenSearch/actions/runs/5791216527/job/15695669089)

### Related Issues
Resolves #9124 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by neetikasinghal** [comment]
@sohami @jed326 please review the changes.

File: CHANGELOG.md [hunk 1]
```diff
@@ -129,4 +130,4 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Security
 
 [Unreleased 3.0]: https://github.com/opensearch-project/OpenSearch/compare/2.x...HEAD
-[Unreleased 2.x]: https://github.com/opensearch-project/OpenSearch/compare/2.10...2.x
```
Comment (jed326) [inline CHANGELOG.md:132]: Try rebasing to see if this goes away

File: CHANGELOG.md [hunk 2]
```diff
@@ -118,6 +118,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Add support for aggregation profiler with concurrent aggregation ([#8801](https://github.com/opensearch-project/OpenSearch/pull/8801))
 - [Remove] Deprecated Fractional ByteSizeValue support #9005 ([#9005](https://github.com/opensearch-project/OpenSearch/pull/9005))
 - Make MultiBucketConsumerService thread safe to use across slices during search ([#9047](https://github.com/opensearch-project/OpenSearch/pull/9047))
+- Rethrow OpenSearch exception in concurrent search ([#9177](https://github.com/opensearch-project/OpenSearch/pull/9177))
```
Comment (jed326) [inline CHANGELOG.md:121]: I think we can make this changelog description more precise -- the issue here is that we wrap `OpenSearchException` in the non-concurrent path with concurrent search.

File: server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java [hunk 3]
```diff
@@ -104,7 +105,7 @@ public AggregationProcessor aggregationProcessor(SearchContext searchContext) {
 
     private static <T extends Exception> void rethrowCauseIfPossible(RuntimeException re, SearchContext searchContext) throws T {
         // Rethrow exception if cause is null
-        if (re.getCause() == null) {
+        if (re.getCause() == null || re instanceof OpenSearchException) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:108]: We surely need a test for it, @neetikasinghal could you please craft one (from the issue)? thank you.

Comment (sohami) [inline server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:108]: why is `re.getCause` null here instead shouldn't it return `OpenSearchException` ? From lucene layer, it seems like it will always throw `RuntimeException` with proper `cause` if a task failed to execute. Ref [here](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java#L65)

Comment (jed326) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:108]: I believe the problem comes from [here](https://github.com/apache/lucene/blob/13e747f95f7b8277cc961e83b4990e6d152dd351/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L665-L668) -- if there is only 1 slice then we just directly call `ContextIndexSearcher::search` and then we directly throw `OpenSearchException`.

Comment (neetikasinghal) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:108]: added the test case for this

Comment (neetikasinghal) [reply server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java:108]: yes @jed326 is right here

File: server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchRethrowExceptionIT.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Collections.singleton(ScriptedExceptionPlugin.class);
+    }
+
+    public void testShardFailureException() {
+        client().prepareIndex("test").setId("1").setSource("field", "value").setRefreshPolicy(IMMEDIATE).get();
+
+        SearchResponse searchResponse = client().prepareSearch("test")
+            .setQuery(
```
Comment (reta) [inline server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchRethrowExceptionIT.java:52]: The test case does the job but is not representative of what is really happening, @jed326 do you think we could simulate the issue by triggering CB during search operation?

Comment (neetikasinghal) [reply server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchRethrowExceptionIT.java:52]: @jed326 thats not true, this test case goes to the `rethrowCauseIfPossible` method and then tests it. So, its not true that we are not testing the code.

@reta yes the issue could be simulated by testing the same set of tests with concurrent search enabled and number of slices=1. I was thinking in this case, I can add all the tests in the `CircuitBreakerServiceIT.java` with concurrent search enabled and the slice count value = 1. Thoughts?

Comment (reta) [reply server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchRethrowExceptionIT.java:52]: >  I was thinking in this case, I can add another class with all the tests in the CircuitBreakerServiceIT.java with concurrent search enabled and the slice count value = 1.

All for it, thank you, as far as we get realistic flow tested

File: server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchCircuitBreakerServiceIT.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import org.opensearch.search.SearchService;
+import org.opensearch.test.OpenSearchIntegTestCase;
+
+import static org.opensearch.test.OpenSearchIntegTestCase.Scope.TEST;
+
+/**
+ * Integration tests for InternalCircuitBreakerService with concurrent search enabled
+ */
+@OpenSearchIntegTestCase.ClusterScope(scope = TEST, numClientNodes = 0, maxNumDataNodes = 1)
+public class ConcurrentSearchCircuitBreakerServiceIT extends CircuitBreakerServiceIT {
```
Comment (reta) [inline server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchCircuitBreakerServiceIT.java:46]: We need https://github.com/opensearch-project/OpenSearch/pull/9083 badly :-)

Comment (neetikasinghal) [reply server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchCircuitBreakerServiceIT.java:46]: yes exactly, but since that is not merged, i couldn't pull that in, I think we can merge that first, then we wouldnt need this extra class.

Comment (jed326) [reply server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchCircuitBreakerServiceIT.java:46]: Apologies if the recommendations are going in circles here but I think we still want to be able to run these tests with multiple slices too? The single slice case is just an edge case we want to test specifically.

**Review by kkmr — APPROVED** [review]

File: server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchCircuitBreakerServiceIT.java (hunk 5 above)
Comment (jed326) [reply server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchCircuitBreakerServiceIT.java:46]: Thinking about this some more -- maybe it is best to just wait until #9083 and then randomize the slice count for this test?

File: server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchCircuitBreakerServiceIT.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/*
```
[the remark this answers is not in the document]
Comment (neetikasinghal) [reply server/src/internalClusterTest/java/org/opensearch/indices/memory/breaker/ConcurrentSearchCircuitBreakerServiceIT.java:9]: I will remove this class as a whole and will add the extension to the parameterized base class.

**Review by reta — APPROVED** [review]

**Review by sohami — APPROVED** [review]
